staging/src/k8s.io/apimachinery/pkg/conversion/conversiontest/apis/v2/types.go

```diff
@@ -0,0 +1,35 @@
+package v2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// To convert from v1 to v2, you have to either default the Type or write logic in your conversion method that effectively
+// does the defaulting to decide what to do.  If you're migrating an older codebase, you need to make sure that nothing
+// in the conversion path relies on having required fields set.  We used to simply allow a failure if that were the case,
+// but if you need to tolerate it for APPLY you have code to inspect.
+// In addition, you need to preserve the pointer aspect of Type through the internal hub to inform the internal->v2 conversion, which didn't use to be the case since defaulting used to happen when going v1->internal.
+// Again, a consideration for migrating an older codebase.
+
+// 1. If a writer sets Reference in v1, do they also set Type in v2?  One piece of information in v1 implies two fields in v2.
```

Comment:
I think the general form of the problem is:

1. There is some set of fields F1 in version A which collectively determine the state of a set of fields F2 in version B, and vice versa.
2. The ordinary case is that the size of F1 is 1 and the size of F2 is also 1 (1:1 correspondence). In this case, it's impossible to have a problem, as either F1 is completely set and implies a setting for F2, or it is absent and F2 should also be absent (and vice versa).
3. I think (without researching) the next most common case is a split: F1 = {Address} and F2 = {Host, Port}. I think this case must either already be doing defaulting during conversion (adding a port to an address), or Port must be an optional field (since Address might have omitted it). (IMO the latter is correct and the former is an API bug.)
4. Finally, since conversion functions are Turing complete, there can be arbitrarily complicated relationships, such as the one you presented here: F1 = {Reference, Type} and F2 = {FooReference, BarReference, Type}

So first, we know that the entire state space of F1 must exactly cover the entire state space of F2, or the types wouldn't be able to pass the round-trip test. Unfortunately, the reason this is true (in your example and likely real world cases) is that the defaulter has arbitrarily removed some states from (in this case) F1 before conversion is invoked. Relying on the defaulter to do this is an API bug, IMO--it is cheating on the round-trip-ability test.

As I see it, there's two options here.
1. Move the unit of defaulting to field groups instead of entire objects; so, if not glolbally defaulting the whole object, we'd still default field `f` iff some other field in a group of which `f` is a member has a value. Under this rule, even in the user's intent, Type would gain a default value.
2. Require "invalid" states to also round-trip. Conversion functions must be bijections. I think this is in practice a new requirement, as we have worked around it up until now.

I suspect we will also have to have a TBD fallback for cases where the existing conversion function can't do 1 or 2 without breaking compatibility.

Reply:
Thanks for the example, I think I now understand this well enough to write a design doc about it.